### PR TITLE
Autoconf

### PR DESCRIPTION
**Pull request by albinahlback**
I will try to list everything that has been done:

## Autoconf
Change from our own configure script to using Autoconf, which is a major improvement. The reason being that more can be tested with fewer lines, and can be generalized more. With this building system I am happy (although some more tests would be nice, and to actually support cross-compilation).

For example, now you can specify include directory and library directory for GMP individually! Amazing, what a feature! Feels like I am in the future!

Moreover, you can specify different prefixes for the installation path, so that the library can be put into `/my/path`, while the headers can go into `/my/other/path`. Truly amazing times we live in.

Also make changes according to `configure.ac`. Now we do not have to manually `cat` into `Makefile`, and we Autoconf can find appropriate programs to compile FLINT.

Also fixed a bug for MinGW when linking the library generated an error due to too long lines. Managed this by creating "merged" object files for each directory (just like before, before my make-makeover, but now also for the static version).

Don't think I mentioned it all, but basically it is now easier to test features during compilation, find suitable programs, search for headers, verify things, and everything in fewer lines of code!

I think I preserved most of the options. I don't think I kept `--with-mpir`, but people using MPIR should know that it is a fork of GMP. I kept a synonym for `--enable-tls`, although I chose to not print a helpstring for it, only for its replacement `--enable-thread-safe`. I chose to change it as that is what MPFR names it. We also do not support `--single` and `--reentrant` anymore as such entries are not compatible with Autoconf. Instead I named it `--enable-reentrant`.

I also like that `config.log` now actually describes the configuration process in detail.

### About shipping configure files
Now, I think that we should always ship configure files in tarballs and so on. However, I do not think they should be present in the repository. For this I think that `make dist` have to be altered, as I suppose that `configure` and `src/config.h.in` has to be in there.

## Remove every `*_DLL`
MinGW did not like it `*_DLL`. Although one could make an exception for MinGW, there is no other system that prefers `*_DLL` (perhaps apart from MSVC). Also bloats the headers.

### CODE REVIEW

**Comment by albinahlback** [comment]
MinGW will probably fail earlier now, and cancel all the other jobs. This is due to changing the building process from CMake to configure+make for MinGW, which is significantly faster. Perhaps I can get a fix for MinGW in the same PR.

**Comment by albinahlback** [comment]
I believe I fixed the problem. Will check that branch first.

**Comment by albinahlback** [comment]
Not sure why Nemo failed. Following the same instructions on my PC works just fine.

**Comment by albinahlback** [comment]
I don't know who is responsible for the code, but it looked like `remove` could be called while the file was associated with an open file stream. I believe that can result in an error.

Anyway, I will now check whether MinGW now works.

The problem with Nemo is problematic, especially since I cannot replicate it.

**Comment by albinahlback** [comment]
Looks like the management of the filestream `qs_inf->siqs` isn't properly managed in `qsieve/large_prime_variant.c` and `qsieve/factor.c`. For example, `large_prime_variant.c` can open the file twice, but there is no other function that starts with checking if the file is already open.

**Comment by albinahlback** [comment]
I have tracked a problem to GR. I will see if I have caused some issue there.

**Comment by albinahlback** [comment]
Okay... So it actually works when using the debugger flag `-g`. It was a bug in the previous `configure` script that if the user explicitly says `CFLAGS=""` it would still override this with the default flags. This is now fixed, so I presume that is why this bug only appears for Nemo.

@fredrik-johansson Do you know what to do? I presume this bug is present in the current code in `trunk`.

**Comment by fredrik-johansson** [comment]
Is there any chance of isolating a minimal .c program that triggers the bug?

**Comment by albinahlback** [comment]
Perhaps the Nemo code fails because of Nemo? If you have a look at the source code for [`factor_squarefree`](https://github.com/Nemocas/Nemo.jl/blob/master/src/flint/fmpz_mod_mpoly.jl#L534), it looks like they have incorrectly written `nmod_mpoly_factor_squarefree` instead of `fmpz_mod_mpoly_factor_squarefree`.

**Comment by albinahlback** [comment]
Anyway, I will try to see if I can give you a breakdown of the failing MinGW code.

**Comment by albinahlback** [comment]
So, perhaps this is totally disjoint from the problem on MinGW, but do the following:
```
./bootstrap.sh
./configure CFLAGS="-g" --enable-assert --disable-static   # and possibly --with-[gmp/mpfr]=blablabal if that's not in your library and include path already
```
Also make sure that everything is uninstalled so nothing gets linked to the wrong library. Then
```
make -j
make -j check
```
and (hopefully) you will see that `fmpz/test/t-is_prime.c` fails, saying that
```
src/fmpz_mod/sub.c:70: _fmpz_mod_subN: Assertion `fmpz_mod_is_canonical(b, ctx)' failed.
```

I traced it back to GR's `_gr_vec_fmpz_mod_vec_sub`, but I will debug with GDB so I am not saying anything wrong here.

**Comment by albinahlback** [comment]
I am a complete newbie at GDB (actually my first real time really using it), and the call stack I believe is:
```
#0  0x00007ffff6aa08ec in ?? () from /usr/lib/libc.so.6
#1  0x00007ffff6a51ea8 in raise () from /usr/lib/libc.so.6
#2  0x00007ffff6a3b53d in abort () from /usr/lib/libc.so.6
#3  0x00007ffff6a3b45c in ?? () from /usr/lib/libc.so.6
#4  0x00007ffff6a4a9f6 in __assert_fail () from /usr/lib/libc.so.6
#5  0x00007ffff6e4dada in _fmpz_mod_subN (a=0x55555562b408, b=0x555555623e38,
    c=0x55555562b408, ctx=0x55555556acb0) at src/fmpz_mod/sub.c:70
#6  0x00007ffff76db6e7 in fmpz_mod_sub (a=0x55555562b408, b=0x555555623e38,
    c=0x55555562b408, ctx=0x55555556acb0) at ./src/fmpz_mod.h:143
#7  0x00007ffff76dbd68 in _gr_fmpz_mod_sub (res=0x55555562b408, x=0x555555623e38,
    y=0x55555562b408, ctx=0x7fffffffdc30) at src/gr/fmpz_mod.c:230
#8  0x00007ffff76eed05 in gr_generic_vec_sub (res=0x55555562b408, src1=0x555555623e38,
    src2=0x55555562b408, len=56, ctx=0x7fffffffdc30) at src/gr/generic.c:2223
#9  0x00007ffff7767582 in _gr_vec_sub (res=0x55555562b408, src1=0x555555623e38,
    src2=0x55555562b408, len=56, ctx=0x7fffffffdc30) at ./src/gr_vec.h:66
#10 0x00007ffff77675d6 in _gr_poly_sub (res=0x55555562b408, poly1=0x555555623e38,
    len1=56, poly2=0x55555562b408, len2=56, ctx=0x7fffffffdc30) at src/gr_poly/sub.c:23
#11 0x00007ffff774a0ec in _gr_poly_div_series_newton (res=0x5555556238e0,
    B=0x555555623c70, Blen=113, A=0x555555623ff8, Alen=113, len=113, cutoff=8,
    ctx=0x7fffffffdc30) at src/gr_poly/div_series_newton.c:78
#12 0x00007ffff7749207 in _gr_poly_div_series_generic (Q=0x5555556238e0,
    A=0x555555623c70, Alen=113, B=0x555555623ff8, Blen=113, len=113, ctx=0x7fffffffdc30)
    at src/gr_poly/div_series.c:29
#13 0x00007ffff7744ed8 in _gr_poly_div_series (res=0x5555556238e0, f=0x555555623c70,
    flen=113, g=0x555555623ff8, glen=113, len=113, ctx=0x7fffffffdc30)
    at ./src/gr_poly.h:212
#14 0x00007ffff774510f in _gr_poly_div_newton (Q=0x5555556238e0, A=0x5555556247d0,
    lenA=241, B=0x555555589c80, lenB=129, ctx=0x7fffffffdc30)
--Type <RET> for more, q to quit, c to continue without paging--
    at src/gr_poly/div_newton.c:55
#15 0x00007ffff7748067 in _gr_poly_divrem_newton (Q=0x5555556238e0, R=0x5555556229d0,
    A=0x5555556247d0, lenA=241, B=0x555555589c80, lenB=129, ctx=0x7fffffffdc30)
    at src/gr_poly/divrem_newton.c:21
#16 0x00007ffff6e60e66 in _fmpz_mod_poly_divrem (Q=0x5555556238e0, R=0x5555556229d0,
    A=0x5555556247d0, lenA=241, B=0x555555589c80, lenB=129, invB=0x7fffffffddb0,
    p=0x7fffffffe200) at src/fmpz_mod_poly/divrem.c:29
#17 0x00007ffff6e62c97 in _fmpz_mod_poly_rem (R=0x5555556229d0, A=0x5555556247d0,
    lenA=241, B=0x555555589c80, lenB=129, invB=0x7fffffffddb0, p=0x7fffffffe200)
    at ./src/fmpz_mod_poly.h:675
#18 0x00007ffff6e62f75 in _fmpz_mod_poly_evaluate_fmpz_vec_fast_precomp (
    vs=0x5555555cf330, poly=0x5555556247d0, plen=241, tree=0x5555555b7990, len=240,
    mod=0x7fffffffe200) at src/fmpz_mod_poly/evaluate_fmpz_vec_fast.c:66
#19 0x00007ffff6fa3e79 in fmpz_factor_pp1 (fac=0x7fffffffe118, n_in=0x7fffffffe200,
    B1=1152, B2sqrt=1050, c=3) at src/fmpz_factor/factor_pp1.c:531
#20 0x00007ffff6dbfff8 in fmpz_is_prime_pocklington (F=0x7fffffffe1e8, R=0x7fffffffe200,
    n=0x7fffffffe290, pm1=0x5555555bdd30, num_pm1=2) at src/fmpz/is_prime_pocklington.c:98
#21 0x00007ffff6dbe7ae in fmpz_is_prime (n=0x7fffffffe290) at src/fmpz/is_prime.c:127
#22 0x000055555555555d in main () at src/fmpz/test/t-is_prime.c:82
```

**Comment by albinahlback** [comment]
As factoring is very vital for primality checks, I will just assume that this is the problem for MinGW.

And we should probably enable assert for at least one CI.

**Comment by fredrik-johansson** [comment]
I think this is another instance of a bug I've seen before. Some basic fmpz_mod_* arithmetic functions *require* input that is reduced modulo N, but there are places in the code where we carelessly construct fmpz_mod_poly or fmpz_mod_mat input that isn't reduced.

Looking at the stack trace, my first guess is that the input to src/fmpz_mod_poly/evaluate_fmpz_vec_fast isn't reduced, but the problem could origin somewhere else.

It would be great to test with asserts at least on some platform to help detect this kind of issue.

**Comment by fredrik-johansson** [comment]
I pushed an immediate fix in 718ee79.

factor_pp1 was doing some fmpz_mod operations, then calling fmpz_poly_mul without reducing the result.

**Comment by albinahlback** [comment]
Alright. Going to rebase. Do you have any feedback or questions regarding the PR?

**Comment by fredrik-johansson** [comment]
I didn't try it out yet.

I have just two comments so far:
* I want to keep the generated ``configure`` in the repo so that the setup is as simple as possible. Besides, it's annoying when there's a difference between the tarball and a clean git checkout.
* We shouldn't break the CMake build. (If CMake can be removed because it's not needed anymore, that's fine of course, but removing it should go in a separate PR.)

**Comment by albinahlback** [comment]
I believe that will screw up the gitignore if someone uses `bootstrap.sh`, but I'll add the appropriate files for it.

And I will see what I can do about CMake.

**Comment by albinahlback** [comment]
Just so you know, it is the generated scripts that stands for like 35k lines of code.

File: config.status
Comment (fingolfin) [inline config.status:1]: Files like `config.status`, `configure`, `config.sub` don't belong into the git repository. They should rather be added to `.gitignore` and be removed from this PR.

**Comment by albinahlback** [comment]
> Yeah, it looks like the generated configure is not going to be portable? (The auxiliary files link to hardcoded paths.) That also means it can't be shipped with tarballs, and autotools will have to be a dependency.

Oh, I forgot that they are links (as `bootstrap.sh` uses `-s` for symbolic links). It is easy to get the raw files, and they are license compatible. However, I still agree with Max in that they should not be present in the repository.

**Comment by albinahlback** [comment]
I removed them.

**Comment by fredrik-johansson** [comment]
How do you propose testing that the configure script placed in a tarball with ``make sdist`` works portably?

**Comment by albinahlback** [comment]
We could make a CI that only activates when pushing tags or something like that!

**Comment by albinahlback** [comment]
Also solves #1267.

**Comment by albinahlback** [comment]
Now the Nemo tests shouldn't fail :)

**Comment by albinahlback** [comment]
Looks good?

**Comment by albinahlback** [comment]
I'll merge this. Feel free to consult me regarding the build system; I consider myself responsible for it now as I have basically changed the entire thing.

**Comment by fredrik-johansson** [comment]
Sure, let's see how it works.

**Comment by fredrik-johansson** [comment]
One bug: ``./configure`` no longer sets ``-g -funroll-loops`` CFLAGS automatically.

**Comment by albinahlback** [comment]
Oh, forgot to say that I removed the flags `-g -funroll-loops` from the default flags. It will set `-mpopcnt` if you haven't specified any `CFLAGS`. If you set your own `CFLAGS`, then `-mpopcnt` shouldn't be added of course.

Do you want me to add `-g -funroll-loops` by default?

**Comment by albinahlback** [comment]
Oh, I see what you are saying. I forgot to set `CFLAGS=$CFLAGS $FLINT_CFLAGS` if user hasn't set `CFLAGS`. Fixing this.

**Comment by fredrik-johansson** [comment]
> It will set -mpopcnt if you haven't specified any CFLAGS.

I don't think it did for me.

> Do you want me to add -g -funroll-loops by default?

Yes.

Determining whether ``-funroll-loops`` can be removed will require profiling.

**Comment by albinahlback** [comment]
In my opinion, we have to check whether `-funroll-loops` will have ANY positive effects at all. Anyway, I opened a PR.

**Pull request merged**